kafka-connect-s3/src/main/java/io/confluent/connect/s3/S3SinkConnectorValidator.java

```diff
@@ -0,0 +1,159 @@
+
+  public static final Map<CompressionType, List<Class<? extends Format>>>
+      COMPRESSION_SUPPORTED_FORMATS = Collections.unmodifiableMap(
+      new HashMap<CompressionType, List<Class<? extends Format>>>() {
+          {
+            put(CompressionType.GZIP, new ArrayList<>(Arrays.asList(
+                    JsonFormat.class,
+                    ByteArrayFormat.class)));
+            put(CompressionType.NONE, new ArrayList<>(Arrays.asList(
+                    AvroFormat.class,
+                    ParquetFormat.class,
+                    JsonFormat.class,
+                    ByteArrayFormat.class)));
+          }
+      }
+    );
+
+  public static final String FORMAT_CONFIG_ERROR_MESSAGE = "Compression Type %s "
+      + "not valid for %s format class: ( %s ).";
+
+  private final Map<String, String> connectorConfigs;
+  private final ConfigDef config;
+
+  protected final Map<String, ConfigValue> valuesByKey;
+
+  public S3SinkConnectorValidator(
+        ConfigDef config, Map<String, String> connectorConfigs, List<ConfigValue> configValues) {
+    this.config = config;
+    this.connectorConfigs = connectorConfigs;
+    valuesByKey = new HashMap<>();
+    for (ConfigValue configValue: configValues) {
+      valuesByKey.put(configValue.name(), configValue);
+    }
+  }
+
+  public Config validate() {
+    log.info("Validating s3 Configs");
+    S3SinkConnectorConfig s3SinkConnectorConfig = null;
+    try {
+      s3SinkConnectorConfig = new S3SinkConnectorConfig(config, connectorConfigs);
+    } catch (ConfigException exception) {
+      log.trace("Configuration not ready for cross validation.");
+    }
+    if (s3SinkConnectorConfig != null) {
+      validateCompression(
+          s3SinkConnectorConfig.getCompressionType(), s3SinkConnectorConfig.formatClass(),
+          s3SinkConnectorConfig.storeKafkaKeys(), s3SinkConnectorConfig.keysFormatClass(),
+          s3SinkConnectorConfig.storeKafkaHeaders(), s3SinkConnectorConfig.headersFormatClass()
+      );
+    }
+
+    return new Config(new ArrayList<>(this.valuesByKey.values()));
+  }
+
+  public void validateCompression(CompressionType compressionType, Class formatClass,
+        boolean storeKafkaKeys, Class keysFormatClass,
+        boolean storeKafkaHeaders, Class headersFormatClass) {
+    List<Class<? extends Format>> validFormats = COMPRESSION_SUPPORTED_FORMATS.get(compressionType);
+    Optional<Class<? extends Format>> formatClassFound = validFormats.stream().filter(
+        format -> format.getName().equals(formatClass.getName()))
+        .findAny();
+    if (!formatClassFound.isPresent()) {
+      recordErrors(
+          String.format(FORMAT_CONFIG_ERROR_MESSAGE,
+              compressionType.name, "data", formatClass.getName()),
+          FORMAT_CLASS_CONFIG, COMPRESSION_TYPE_CONFIG);
+    }
+
+    if (storeKafkaKeys) {
+      Optional<Class<? extends Format>> keysFormatClassFound = validFormats.stream().filter(
```

Comment:
if validFormats inside the map is modelled as set instead of list, then we can avoid creating a stream, filter and find